<a href="https://cognitiveclass.ai/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDS0701ESCoursera21685115-2022-01-01"><img src = "https://ibm.box.com/shared/static/9gegpsmnsoo25ikkbl4qzlvlyjbgxs5x.png" width = 400> </a>

<h1 align=center><font size = 5>Cómo usar la API de FourSquare con Python</font></h1>


## Introducción

En este laboratorio usted aprenderá detalladamente a realizar llamadas a la API de FourSquare con distintos propositos. Aprenderá a construir una URL para enviar peticiones a la API para buscar lugares específicos, explorar un sitio en particular, un usuario de FourSquare o alguna referencia geografica, además, podrá obtener los lugares de moda alrededor de alguna locación. También aprederá a usar la librería de visualización Folium para ver los resultados.


## Indice

1.  <a href="https://#item1">Función de Busqueda de la API de Foursquare</a>
2.  <a href="https://#item2">Explorar un Lugar</a>
3.  <a href="https://#item3">Explorar un Usuario</a>
4.  <a href="https://#item4">Función de Exploración de la API de Foursquare</a>
5.  <a href="https://#item5">Obtener Lugares de Moda</a>


### Importar las librerías necesarias


In [1]:
import requests # librería para manejar las solicitudes
import pandas as pd # librería para análisis de datos
import numpy as np # librería para manejar datos vectorizados
import random # librería para generar números aleatorios


!pip install geopy
from geopy.geocoders import Nominatim # módulo para convertir una dirección en valores de latitud y longitud 

# librerías para mostrar imágenes 
from IPython.display import Image 
from IPython.core.display import HTML 
    
# librería para convertir un archivo json en un dataframe pandas
from pandas.io.json import json_normalize


! pip install folium==0.5.0
import folium # librería para graficar 

print('Folium installed')
print('Libraries imported.')

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 119.8/119.8 kB 11.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.3/40.3 kB 4.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.2/79.2 kB 8.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for folium: filename=folium-0.5.0-py3-none-any.whl size=76117 sha256=1b6e1c6108ff9373d0c137a259064000112bd7c6d1d0f103a094a0e6a04bfee5
  Stored in directory: /home/jupyterlab/.cache/pip/wheels/26/e2/90/e29b0a163b7ca7737dd33fc1ad742b38ff55963d8220013134
Successfully built folium
  Attempting uninstall: folium
    Found existing installation: folium 0.11.0
    Uninstalling folium-0.11.0:
      Successfully uninstalled folium-0.11.0
Folium installed
Libraries imported.


### Definir la Versión y Credenciales de FourSquare


##### Asegúrese de haber creado una cuenta de desarrollador en FourSquare con las credenciales apropiadas


In [2]:
CLIENT_ID = '55S2ZJJJVK3T1D1ZDNQLHEAH15HG4AHHMYJZNETKJK5EXKBB' # su ID de Foursquare
CLIENT_SECRET = 'NRH3ODLT453CFZY2C0P33DJEDKCXROHBNC1OQSMT52ST3TD4' # su Secreto de Cliente de Foursquare
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 55S2ZJJJVK3T1D1ZDNQLHEAH15HG4AHHMYJZNETKJK5EXKBB
CLIENT_SECRET:NRH3ODLT453CFZY2C0P33DJEDKCXROHBNC1OQSMT52ST3TD4


#### Vamos a suponer que usted esta hospedado en el hotel Conrad. Empecemos por convertir la dirección del hotel en valores de coordenadas dadas por la latitud y la longitud.


Para definir una instancia del geocoder necesitamos definir un user_agent. Nombraremos a nuestro agente <em>foursquare_agent</em> de la siguiente forma.


In [3]:
address = '102 North End Ave, New York, NY'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print(latitude, longitude)

40.7149555 -74.0153365


<a id="item1"></a>


## 1. Búsqueda por categoría de un lugar específico

> `https://api.foursquare.com/v2/venues/`**search**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&ll=`**LATITUDE**`,`**LONGITUDE**`&v=`**VERSION**`&query=`**QUERY**`&radius=`**RADIUS**`&limit=`**LIMIT**


#### Supongamos que es la hora de comer y usted desea comida italiana. Vamos a definir una consulta para buscar comida italiana dentro de un radio de 500 metros del hotel Conrad.


In [4]:
search_query = 'Italian'
radius = 500
print(search_query + ' .... OK!')

Italian .... OK!


In [11]:
import requests

search_query = 'Italian'
radius = 500
print(search_query + ' .... OK!')

#### Definir la URL correspondiente


In [25]:
#url = "https://api.foursquare.com/v3/places/search?client_id=55S2ZJJJVK3T1D1ZDNQLHEAH15HG4AHHMYJZNETKJK5EXKBB&client_secret=NRH3ODLT453CFZY2C0P33DJEDKCXROHBNC1OQSMT52ST3TD4&v=20180602&||=40.73,-74.01&query=coffee"
url = 'https://api.foursquare.com/v3/places/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
headers = {
    "Accept": "application/json",
    "Authorization": "fsq3a5oXiMEiuZYusTTUUGz8CHLfGZvmouDBQSuiyW4QVys="
}
url



'https://api.foursquare.com/v3/places/search?client_id=55S2ZJJJVK3T1D1ZDNQLHEAH15HG4AHHMYJZNETKJK5EXKBB&client_secret=NRH3ODLT453CFZY2C0P33DJEDKCXROHBNC1OQSMT52ST3TD4&ll=40.7149555,-74.0153365&v=20180604&query=Italian&radius=500&limit=30'

#### Enviar la petición GET y examinar los resultados


In [26]:
results = requests.request("GET", url, headers=headers).json()
results

{'results': [{'fsq_id': '4fa862b3e4b0ebff2f749f06',
   'categories': [{'id': 13064,
     'name': 'Pizzeria',
     'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/pizza_',
      'suffix': '.png'}},
    {'id': 13236,
     'name': 'Italian Restaurant',
     'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/italian_',
      'suffix': '.png'}}],
   'chains': [],
   'distance': 58,
   'geocodes': {'main': {'latitude': 40.715256, 'longitude': -74.014835},
    'roof': {'latitude': 40.715256, 'longitude': -74.014835}},
   'link': '/v3/places/4fa862b3e4b0ebff2f749f06',
   'location': {'address': '225 Murray St',
    'census_block': '360610317033007',
    'country': 'US',
    'dma': 'New York',
    'formatted_address': '225 Murray St, New York, NY 10282',
    'locality': 'New York',
    'postcode': '10282',
    'region': 'NY'},
   'name': "Harry's Italian Pizza Bar",
   'related_places': {},
   'timezone': 'America/New_York'},
  {'fsq_id': '5d41d093104e90000882db55',

#### Obtener la parte relevante del JSON y convertirlo en un dataframe *pandas*


In [32]:
# asignar la parte relevante del JSON a la variable venues 
venues = results['results']

# convertir venues en un dataframe
dataframe = json_normalize(venues)
dataframe.head()

/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/ipykernel_launcher.py:5: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  """


fsq_id  \
0  4fa862b3e4b0ebff2f749f06   
1  5d41d093104e90000882db55   
2  556cb8c6498e751e51fbcf41   
3  3fd66200f964a520aae61ee3   
4  592b5e9c0ff4f9137e2a8390   

                                          categories chains  distance  \
0  [{'id': 13064, 'name': 'Pizzeria', 'icon': {'p...     []        58   
1  [{'id': 13236, 'name': 'Italian Restaurant', '...     []       170   
2  [{'id': 13236, 'name': 'Italian Restaurant', '...     []        92   
3  [{'id': 13065, 'name': 'Restaurant', 'icon': {...     []       470   
4  [{'id': 13025, 'name': 'Wine Bar', 'icon': {'p...     []       490   

                                  link                       name  \
0  /v3/places/4fa862b3e4b0ebff2f749f06  Harry's Italian Pizza Bar   
1  /v3/places/5d41d093104e90000882db55              Sant Ambroeus   
2  /v3/places/556cb8c6498e751e51fbcf41                       Parm   
3  /v3/places/3fd66200f964a520aae61ee3           Gigino Trattoria   
4  /v3/places/592b5e9c0ff4f9137e2a8390          L' Angolo Tribeca   

           timezone  geocodes.main.latitude  geocodes.main.longitude  \
0  America/New_York               40.715256               -74.014835   
1  America/New_York               40.713759               -74.014339   
2  America/New_York               40.714470               -74.015939   
3  America/New_York               40.717261               -74.010506   
4  America/New_York               40.717209               -74.010461   

   geocodes.roof.latitude  ...  location.locality location.postcode  \
0               40.715256  ...           New York             10282   
1               40.713759  ...           New York             10007   
2               40.714470  ...           New York             10080   
3               40.717261  ...           New York             10013   
4               40.717209  ...           New York             10013   

  location.region   location.cross_street related_places.parent.fsq_id  \
0              NY                     NaN                          NaN   
1              NY                             4a6fa0abf964a520a4d61fe3   
2              NY                 West St                          NaN   
3              NY  btw Duane & Reade Sts.                          NaN   
4              NY               Greenwich                          NaN   

  related_places.parent.name geocodes.drop_off.latitude  \
0                        NaN                        NaN   
1           Brookfield Place                        NaN   
2                        NaN                        NaN   
3                        NaN                  40.717301   
4                        NaN                  40.717441   

  geocodes.drop_off.longitude location.address_extended  \
0                         NaN                       NaN   
1                         NaN                       NaN   
2                         NaN                       NaN   
3                  -74.010671                    Frnt 1   
4                  -74.010434                       NaN   

  related_places.children  
0                     NaN  
1                     NaN  
2                     NaN  
3                     NaN  
4                     NaN  

[5 rows x 26 columns]

#### Definir la información de interes y filtrar el dataframe


In [38]:
# mantener unicamente las columnas que incluyan el nombre del lugar y cualquier cosa asociada a la ubicación
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['fsq_id'] + [col for col in dataframe.columns if col.startswith('geocodes.main')]
dataframe_filtered = dataframe.loc[:, filtered_columns]

# función para extraer la categoría del lugar
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filtrar la categoría para cada fila
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# limpiar la columna "names" manteniendo solo el último término
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered


name                   categories  \
0         Harry's Italian Pizza Bar                     Pizzeria   
1                     Sant Ambroeus           Italian Restaurant   
2                              Parm           Italian Restaurant   
3                  Gigino Trattoria                   Restaurant   
4                 L' Angolo Tribeca                     Wine Bar   
5                    Cicci Di Carne           Italian Restaurant   
6                    Sauce pizzeria                     Pizzeria   
7                       Shake Shack                 Burger Joint   
8                          iPizzaNY           Italian Restaurant   
9                  Brookfield Place                Shopping Mall   
10           Amore Pizza Restaurant                     Pizzeria   
11                       Blue Smoke                    BBQ Joint   
12                           El Vez          American Restaurant   
13                      Hudson Eats                   Food Court   
14                            Atrio          American Restaurant   
15                      Whole Foods  Grocery Store / Supermarket   
16                   Global Cuisine           Chinese Restaurant   
17             Winter Garden Atrium        Performing Arts Venue   
18                              Dig          American Restaurant   
19  Zucker's Bagels and Smoked Fish                   Bagel Shop   
20              Del Frisco's Grille                   Steakhouse   
21             Windows on the World          American Restaurant   
22                           Eataly           Italian Restaurant   
23                    P.J. Clarke's                 Burger Joint   
24                   Gee Whiz Diner                 Burger Joint   
25            Conca Cucina Italiana                     Pizzeria   

                                              address     census_block  \
0                                       225 Murray St  360610317033007   
1                                                 NaN  360610317031002   
2                                        250 Vesey St  360610317031002   
3                                    323 Greenwich St  360610039003010   
4                                       190A Duane St  360610039003010   
5                                        250 Vesey St  360610317031002   
6                                      225 Liberty St  360610317031002   
7                                       215 Murray St  360610317033007   
8                                     147 Chambers St  360610021003000   
9                                        230 Vesey St  360610317031002   
10                                    147 Chambers St  360610021003000   
11                                       255 Vesey St  360610317033007   
12                                       259 Vesey St  360610317033007   
13                                     225 Liberty St  360610317031002   
14                                      102 N End Ave  360610317033007   
15                                   270 Greenwich St  360610021004001   
16                                      222 Murray St  360610317033007   
17  World Financial Center - Belford/Harbor Way, N...  360610317031002   
18                                     225 Liberty St  360610317031002   
19                                    146 Chambers St  360610021003001   
20                                       250 Vesey St  360610317031002   
21                                  1 World Trade Ctr  360610013005000   
22                                   185 Greenwich St  360610013005000   
23                                     200 Liberty St  360610317031002   
24                                   295 Greenwich St  360610021003001   
25                                      63 W Broadway  360610021001006   

   country       dma                                  formatted_address  \
0       US  New York                  225 Murray St, New York, NY 10282   
1       US  New York                         

#### Visualicemos los restaurantes italianos cercanos


In [50]:
dataframe_filtered.name

0           Harry's Italian Pizza Bar
1                       Sant Ambroeus
2                                Parm
3                    Gigino Trattoria
4                   L' Angolo Tribeca
5                      Cicci Di Carne
6                      Sauce pizzeria
7                         Shake Shack
8                            iPizzaNY
9                    Brookfield Place
10             Amore Pizza Restaurant
11                         Blue Smoke
12                             El Vez
13                        Hudson Eats
14                              Atrio
15                        Whole Foods
16                     Global Cuisine
17               Winter Garden Atrium
18                                Dig
19    Zucker's Bagels and Smoked Fish
20                Del Frisco's Grille
21               Windows on the World
22                             Eataly
23                      P.J. Clarke's
24                     Gee Whiz Diner
25              Conca Cucina Italiana
Name: name, 

In [55]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=13) # generate map centred around the Conrad Hotel

# añadir un circulo rojo que represente el Hotel Conrad
folium.CircleMarker(
    [latitude, longitude],
    radius=10,
    color='red',
    popup='Conrad Hotel',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(venues_map)

# añadir los restaurantes italianos como circulo azules
for lat, lng, label in zip(dataframe_filtered.latitude, dataframe_filtered.longitude, dataframe_filtered.categories):
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

# mostrar el mapa
venues_map

<a id="item2"></a>


## 2. Explorar un Lugar Dado

> `https://api.foursquare.com/v2/venues/`**VENUE_ID**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&v=`**VERSION**


### A. Investiguemos el restaurante italiano mas cercano -- *Harry's Italian Pizza Bar*


In [56]:
venue_id = '4fa862b3e4b0ebff2f749f06' # ID of Harry's Italian Pizza Bar
url = 'https://api.foursquare.com/v3/places/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)
url

'https://api.foursquare.com/v3/places/4fa862b3e4b0ebff2f749f06?client_id=55S2ZJJJVK3T1D1ZDNQLHEAH15HG4AHHMYJZNETKJK5EXKBB&client_secret=NRH3ODLT453CFZY2C0P33DJEDKCXROHBNC1OQSMT52ST3TD4&v=20180604'

#### Enviar la petición GET y examinar los resultados


In [59]:
result = requests.request("GET", url, headers=headers).json()
print(result.keys())
result


dict_keys(['fsq_id', 'categories', 'chains', 'geocodes', 'link', 'location', 'name', 'related_places', 'timezone'])


{'fsq_id': '4fa862b3e4b0ebff2f749f06',
 'categories': [{'id': 13064,
   'name': 'Pizzeria',
   'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/pizza_',
    'suffix': '.png'}},
  {'id': 13236,
   'name': 'Italian Restaurant',
   'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/italian_',
    'suffix': '.png'}}],
 'chains': [],
 'geocodes': {'main': {'latitude': 40.715256, 'longitude': -74.014835},
  'roof': {'latitude': 40.715256, 'longitude': -74.014835}},
 'link': '/v3/places/4fa862b3e4b0ebff2f749f06',
 'location': {'address': '225 Murray St',
  'census_block': '360610317033007',
  'country': 'US',
  'dma': 'New York',
  'formatted_address': '225 Murray St, New York, NY 10282',
  'locality': 'New York',
  'postcode': '10282',
  'region': 'NY'},
 'name': "Harry's Italian Pizza Bar",
 'related_places': {},
 'timezone': 'America/New_York'}

### B. Obtener la puntuación general del lugar


In [60]:
try:
    print(result['rating'])
except:
    print('This venue has not been rated yet.')

This venue has not been rated yet.


No es una puntuación muy buena. Revisemos la puntuación del siguiente restaurante italiano mas cercano.


In [61]:
venue_id = '4f3232e219836c91c7bfde94' # ID del Restaurante Conca Cucina Italian
url = 'https://api.foursquare.com/v3/places/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

result = requests.request("GET", url, headers=headers).json()
try:
    print(result['rating'])
except:
    print('This venue has not been rated yet.')

This venue has not been rated yet.


Debido a que este restaurante no tiene puntuación, revisemos un tercero


In [68]:
venue_id = '3fd66200f964a520f4e41ee3' # ID de Ecco
url = 'https://api.foursquare.com/v3/places/{}/tips'.format(venue_id)

result_tips = requests.request("GET", url, headers=headers).json()

result_tips


[{'id': '5ab1cb46c9a517174651d3fe',
  'created_at': '2018-03-21T03:02:30.000Z',
  'text': 'A+ Italian food! Trust me on this: my mom’s side of the family is 100% Italian. I was born and bred to know good pasta when I see it, and Ecco is one of my all-time NYC favorites'},
 {'id': '5cb7051b180b910039f90625',
  'created_at': '2019-04-17T10:51:07.000Z',
  'text': 'Excellent food!! Osso bucco special one of the best I ever had! Lobster ravioli with porcini mushroomhomemade Italian cheesecake, tiramisu and napoleons...calamari fra diavolo was sautéed not fried'},
 {'id': '5247cb68498eeac13eb4c3e1',
  'created_at': '2013-09-29T06:40:40.000Z',
  'text': 'Service was truly excellent- attentive, quick, & friendly. Food was equally amazing- particularly calamari and pollo alla scarpiello. The sauces were incredibly flavorful. Great wine list too! A+++'},
 {'id': '5bc9fa3f5455b2002c84f781',
  'created_at': '2018-10-19T15:37:35.000Z',
  'text': 'Ignore the Wall Street types, you are here for the f

In [69]:
venue_id = '3fd66200f964a520f4e41ee3' # ID de Ecco
url = 'https://api.foursquare.com/v3/places/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

result = requests.request("GET", url, headers=headers).json()
try:
    print(result['rating'])
except:
    print('This venue has not been rated yet.')

result.keys()
result

This venue has not been rated yet.


{'fsq_id': '3fd66200f964a520f4e41ee3',
 'categories': [{'id': 13236,
   'name': 'Italian Restaurant',
   'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/italian_',
    'suffix': '.png'}}],
 'chains': [],
 'geocodes': {'main': {'latitude': 40.715192, 'longitude': -74.008808}},
 'link': '/v3/places/3fd66200f964a520f4e41ee3',
 'location': {'address': '124 Chambers St',
  'address_extended': 'Apt 2',
  'census_block': '360610021001002',
  'country': 'US',
  'cross_street': 'btwn Church St & W Broadway',
  'dma': 'New York',
  'formatted_address': '124 Chambers St (btwn Church St & W Broadway), New York, NY 10007',
  'locality': 'New York',
  'postcode': '10007',
  'region': 'NY'},
 'name': 'Ecco',
 'related_places': {},
 'timezone': 'America/New_York'}

Este restaurante tiene un puntuación ligeramente mejor, revisemosla con mas detalle.


### C. Obtener el número de consejos


In [73]:
result_tips

[{'id': '5ab1cb46c9a517174651d3fe',
  'created_at': '2018-03-21T03:02:30.000Z',
  'text': 'A+ Italian food! Trust me on this: my mom’s side of the family is 100% Italian. I was born and bred to know good pasta when I see it, and Ecco is one of my all-time NYC favorites'},
 {'id': '5cb7051b180b910039f90625',
  'created_at': '2019-04-17T10:51:07.000Z',
  'text': 'Excellent food!! Osso bucco special one of the best I ever had! Lobster ravioli with porcini mushroomhomemade Italian cheesecake, tiramisu and napoleons...calamari fra diavolo was sautéed not fried'},
 {'id': '5247cb68498eeac13eb4c3e1',
  'created_at': '2013-09-29T06:40:40.000Z',
  'text': 'Service was truly excellent- attentive, quick, & friendly. Food was equally amazing- particularly calamari and pollo alla scarpiello. The sauces were incredibly flavorful. Great wine list too! A+++'},
 {'id': '5bc9fa3f5455b2002c84f781',
  'created_at': '2018-10-19T15:37:35.000Z',
  'text': 'Ignore the Wall Street types, you are here for the f

### D. Obtener los consejos de un lugar

> `https://api.foursquare.com/v2/venues/`**VENUE_ID**`/tips?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&v=`**VERSION**`&limit=`**LIMIT**


#### Generar una URL y enviar una solicitud GET. Asegúrese de establecer el límite para obtener todos los consejos


In [74]:
## Consejos de Ecco
limit = 15 # establece el límite para ser mayor o igual al número total de consejos
url = 'https://api.foursquare.com/v3/places/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION, limit)

results = requests.request("GET", url, headers=headers).json()
results

[{'id': '5ab1cb46c9a517174651d3fe',
  'created_at': '2018-03-21T03:02:30.000Z',
  'text': 'A+ Italian food! Trust me on this: my mom’s side of the family is 100% Italian. I was born and bred to know good pasta when I see it, and Ecco is one of my all-time NYC favorites'},
 {'id': '5cb7051b180b910039f90625',
  'created_at': '2019-04-17T10:51:07.000Z',
  'text': 'Excellent food!! Osso bucco special one of the best I ever had! Lobster ravioli with porcini mushroomhomemade Italian cheesecake, tiramisu and napoleons...calamari fra diavolo was sautéed not fried'},
 {'id': '5247cb68498eeac13eb4c3e1',
  'created_at': '2013-09-29T06:40:40.000Z',
  'text': 'Service was truly excellent- attentive, quick, & friendly. Food was equally amazing- particularly calamari and pollo alla scarpiello. The sauces were incredibly flavorful. Great wine list too! A+++'},
 {'id': '5bc9fa3f5455b2002c84f781',
  'created_at': '2018-10-19T15:37:35.000Z',
  'text': 'Ignore the Wall Street types, you are here for the f

#### Obtener consejos y la lista de características asociadas


In [75]:
tips = results

tip = results[0]
tip.keys()

dict_keys(['id', 'created_at', 'text'])

#### Dar formato al ancho de la columna y mostrar todos los consejos


In [76]:
pd.set_option('display.max_colwidth', -1)

tips_df = json_normalize(tips) # consejos json normalizados

# columnas a mantener
#filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.lastName', 'user.id']
filtered_columns = ['text', 'id', 'created_at']
tips_filtered = tips_df.loc[:, filtered_columns]

# mostrar consejos
tips_filtered.reindex()

/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  """Entry point for launching an IPython kernel.
/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


text  \
0   A+ Italian food! Trust me on this: my mom’s side of the family is 100% Italian. I was born and bred to know good pasta when I see it, and Ecco is one of my all-time NYC favorites                      
1   Excellent food!! Osso bucco special one of the best I ever had! Lobster ravioli with porcini mushroomhomemade Italian cheesecake, tiramisu and napoleons...calamari fra diavolo was sautéed not fried   
2   Service was truly excellent- attentive, quick, & friendly. Food was equally amazing- particularly calamari and pollo alla scarpiello. The sauces were incredibly flavorful. Great wine list too! A+++   
3   Ignore the Wall Street types, you are here for the food. Baked clams, gnocchi, house veal. Old school New York Italian in a gorgeous saloon style space.                                                
4   Old world comfort where you can enjoy veal chop alla salvia, lobster fra diavolo and branzino al forno. You can get our free Downtown dining guide at http://www.downtownny.com/mapsguides/             
5   Everything is amazing in this place.                                                                                                                                                                    
6   Alana is awesome! She makes you feel at home no matter how far you have flown.                                                                                                                          
7   New York strip to die for                                                                                                                                                                               
8   Sadly, Ecco is shuttered after more than 35 years in TriBeCa                                                                                                                                            
9   Sadly, Ecco’s last day after decades in Chambers was December 31, 2019. New landlord, untenable rent increase.                                                                                          
10  Service. Food and drinks. Love it 100%👍👍👍                                                                                                                                                               
11  It was ok! Nothing excepcional, but you can say that is a honest service.                                                                                                                               
12  lobster ravioli all the way but most importantly, pay attention to their specials..                                                                                                                     
13  The food and service is excellent.                                                                                                                                                                      
14  Veal Milanese is classic and very good.                                                                                                                                                                 

                          id                created_at  
0   5ab1cb46c9a517174651d3fe  2018-03-21T03:02:30.000Z  
1   5cb7051b180b910039f90625  2019-04-17T10:51:07.000Z  
2   5247cb68498eeac13eb4c3e1  2013-09-29T06:40:40.000Z  
3   5bc9fa3f5455b2002c84f781  2018-10-19T15:37:35.000Z  
4   4c7ac41f278eb713ae245c80  2010-08-29T20:33:35.000Z  
5   5a9a1dc70d173f60ff04a4e7  2018-03-03T04:00:07.000Z  
6   4f289fdfe4b0adf39f89a832  2012-02-01T02:13:51.000Z  
7   563e0df7cd100e587973828a  2015-11-07T14:43:03.000Z  
8   5e1e7410ccf4480006a2701d  2020-01-15T02:08:16.000Z  
9   5e0c9e6fa1849300078365c1  2020-01-01T13:28:15.000Z  
10  53e54fd6498ef3c011579655  2014-08-08T22:31:50.000Z  
11  528978d811d2f94177d1fcdc  2013-11-18T02:18:00.000Z  
12  4ca4e1fa965c9c744604c7fa  2010-09-30T19:16:10.000Z  
13  52f7977a11d262c1858cc523  2014-02-09T14:58:02.000Z  
14  4bce651c70c603bbf29b98b4  2010-04-21T02:38:20.000Z

Recuerde que debido a que estamos usando una cuenta personal de desarrollador solo podemos acceder a 2 de los 15 consejos del restaurante.


<a id="item3"></a>


## 3. Buscar a un Usuario de FourSquare

> `https://api.foursquare.com/v2/users/`**USER_ID**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&v=`**VERSION**


### Definir la URL, enviar la solicitud GET y mostrar las características asociadas con el usuario


In [ ]:
idnumber = '48454224' # ID del usuario con mayor número de aceptación y perfil completo 

url = 'https://api.foursquare.com/v2/users/{}/tips?client_id={}&client_secret={}&v={}'.format(idnumber, CLIENT_ID, CLIENT_SECRET, VERSION) # define URL

# enviar solicitud GET
results = requests.get(url).json()

#user_data = results['response']['user']
user_data=results['response']['tips']['items'][0]['venue']['photos']['groups'][0]['items']#['items']
# mostrar características asociadas con el usuario
#user_data.keys()
#results
pd.set_option('display.max_colwidth', -1)

users_df = json_normalize(user_data)

# filtrar columnas
filtered_columns = ['id','user.id','user.firstName','user.lastName']
tips_filtered = users_df.loc[:, filtered_columns]

# mostrar consejos del usuario
df=tips_filtered
df
#url

In [ ]:

g=df.loc[df['user.id'] == '133773133']
print('First Name: ' + g['user.firstName'])
print('Last Name: ' + g['user.lastName'])


### Obtener la Imagen del Perfil del Usuario


In [ ]:
# 1. tomar el prefijo de la foto
# 2. tomar el sufijo de la foto
# 3. concatenarlos utilizando el tamaño de la imagen
Image(url='https://fastly.4sqi.net/img/general/540x920/133773133_ODR5Au05ENkSyu3xxfV3VOfCa0idDfm9Q4n8YbhZDRQ.jpg')

¡Wow! Parece ser que Nick es un usuario muy activo en FourSquare con mas de 250 consejos.


### Obtener las Recomendaciones del usuario


In [78]:
# definir la URL de las recomendaciones
user_id='484542633'
url = 'https://api.foursquare.com/v3/users/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format(user_id, CLIENT_ID, CLIENT_SECRET, VERSION, limit)

# enviar la solicitud GET y obtener las recomendaciones del usuario
results = requests.request("GET", url, headers=headers).json()
tips = results
tips

{'message': "Endpoint '/v3/users/484542633/tips' not found."}

In [77]:
# definir la URL de las recomendaciones
user_id='484542633'
url = 'https://api.foursquare.com/v3/users/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format(user_id, CLIENT_ID, CLIENT_SECRET, VERSION, limit)

# enviar la solicitud GET y obtener las recomendaciones del usuario
results = requests.request("GET", url, headers=headers).json()
tips = results['response']['tips']['items']

# dar formato al ancho de la columna
pd.set_option('display.max_colwidth', -1)

tips_df = json_normalize(tips)

# filtrar columnas
filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id']
tips_filtered = tips_df.loc[:, filtered_columns]

# mostrar las recomendaciones del usuario
tips_filtered

KeyError: 'response'

#### Obtengamos el lugar para la recomendación con el mayor número de aceptaciones


In [ ]:
tip_id = '5ab5575d73fe2516ad8f363b' # id de la recomendación

# definir la URL
url = 'http://api.foursquare.com/v2/tips/{}?client_id={}&client_secret={}&v={}'.format(tip_id, CLIENT_ID, CLIENT_SECRET, VERSION)

# enviar la solicitud GET y examinar los resultados
result = requests.get(url).json()
print(result['response']['tip']['venue']['name'])
print(result['response']['tip']['venue']['location'])

## 4. Explorar una Ubicación

> `https://api.foursquare.com/v2/venues/`**explore**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&ll=`**LATITUDE**`,`**LONGITUDE**`&v=`**VERSION**`&limit=`**LIMIT**


#### Ha definido su platillo gourmet en Ecco y siente curiosidad acerca de los sitios populares alrededor del restaurante. Para explorar el área, empecemos por obtener los valores de latitud y la longitud del Restaurant Ecco.


In [79]:
latitude = 40.715337
longitude = -74.008848

#### Definir una URL


In [84]:
url = 'https://api.foursquare.com/v3/places/search?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v3/places/search?client_id=55S2ZJJJVK3T1D1ZDNQLHEAH15HG4AHHMYJZNETKJK5EXKBB&client_secret=NRH3ODLT453CFZY2C0P33DJEDKCXROHBNC1OQSMT52ST3TD4&ll=40.715337,-74.008848&v=20180604&radius=500&limit=30'

#### Enviar una solicitud GET y examinar los resultados


In [81]:
import requests

In [86]:
results = requests.request("GET", url, headers=headers).json()
results.head()
#'There are {} around Ecco restaurant.'.format(len(results['response']['groups'][0]['items']))

AttributeError: 'dict' object has no attribute 'head'

#### Obtener la parte relevante del objeto JSON


In [ ]:
items = results['response']['groups'][0]['items']
items[0]

#### Procesar el objeto JSON, convetirlo y limpiar el dataframe


In [ ]:
dataframe = json_normalize(items) # JSON

# filtrar columnas
filtered_columns = ['venue.name', 'venue.categories'] + [col for col in dataframe.columns if col.startswith('venue.location.')] + ['venue.id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# filtrar la categoría para cada columna
dataframe_filtered['venue.categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# limpiar columnas
dataframe_filtered.columns = [col.split('.')[-1] for col in dataframe_filtered.columns]

dataframe_filtered.head(10)

#### Visualicemos estos elementos en el mapa alrededor de nuestra ubicación


In [ ]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=15) # generate map centred around Ecco


# Añadir Ecco como una marca circular roja
folium.CircleMarker(
    [latitude, longitude],
    radius=10,
    popup='Ecco',
    fill=True,
    color='red',
    fill_color='red',
    fill_opacity=0.6
    ).add_to(venues_map)


# añadir sitios populares al mapa como marcas circulares azules
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        fill=True,
        color='blue',
        fill_color='blue',
        fill_opacity=0.6
        ).add_to(venues_map)

# mostrar mapa
venues_map

<a id="item5"></a>


## 5. Explorar Sitios Populares

> `https://api.foursquare.com/v2/venues/`**trending**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&ll=`**LATITUDE**`,`**LONGITUDE**`&v=`**VERSION**


#### Ahora, en vez de explorar  el área alrededor de Ecco, es de interes saber los sitios que son populares durante su comida, esto es, los lugares con el mayor tráfico peatonal. Hagamoslo y obtengamos los sitios populares alrededor de Ecco.


In [ ]:
# definir la URL
url = 'https://api.foursquare.com/v2/venues/trending?client_id={}&client_secret={}&ll={},{}&v={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION)

# enviar la solicitud GET y obtener los sitios de moda
results = requests.get(url).json()
results

### Revisar si algún sitio esta de moda en este momento


In [ ]:
if len(results['response']['venues']) == 0:
    trending_venues_df = 'No trending venues are available at the moment!'
    
else:
    trending_venues = results['response']['venues']
    trending_venues_df = json_normalize(trending_venues)

    # filtrar columnas
    columns_filtered = ['name', 'categories'] + ['location.distance', 'location.city', 'location.postalCode', 'location.state', 'location.country', 'location.lat', 'location.lng']
    trending_venues_df = trending_venues_df.loc[:, columns_filtered]

    # filtrar la categoría para cada fila
    trending_venues_df['categories'] = trending_venues_df.apply(get_category_type, axis=1)

In [ ]:
# mostrar los sitios populares
trending_venues_df

Ahora, dependiendo de cuando usted ejecute el codigo anterior, podría obtener distintos sitios ya que la información de tráfico peatonal se obtiene en tiempo real.


### Visuaizar sitios populares


In [ ]:
if len(results['response']['venues']) == 0:
    venues_map = 'Cannot generate visual as no trending venues are available at the moment!'

else:
    venues_map = folium.Map(location=[latitude, longitude], zoom_start=15) # generate map centred around Ecco


    # añadir Ecco como una marca circular roja
    folium.CircleMarker(
        [latitude, longitude],
        radius=10,
        popup='Ecco',
        fill=True,
        color='red',
        fill_color='red',
        fill_opacity=0.6
    ).add_to(venues_map)


    # añadir los sitios populares como marcas circulares azules
    for lat, lng, label in zip(trending_venues_df['location.lat'], trending_venues_df['location.lng'], trending_venues_df['name']):
        folium.CircleMarker(
            [lat, lng],
            radius=5,
            poup=label,
            fill=True,
            color='blue',
            fill_color='blue',
            fill_opacity=0.6
        ).add_to(venues_map)

In [ ]:
# mostrar el mapa
venues_map

<a id="item6"></a>


### Gracias por terminar este laboratorio

Este cuaderno fue creado por [Alex Aklson](https://www.linkedin.com/in/aklson?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDS0701ESCoursera21685115-2022-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork-21253531&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ). Espero que haya encontrado este laboratorio de su interes y educativo. Tómese la libertad de contactar conmigo para cualquier duda o aclaración.

Este cuaderno fue modificado por Nayef Abou Tayoun ([https://www.linkedin.com/in/nayefaboutayoun/](https://www.linkedin.com/in/nayefaboutayoun?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDS0701ESCoursera21685115-2022-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork-21253531&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ))


Este cuaderno forma parte del curso en **Coursera** llamado *Applied Data Science Capstone*. Si accede a este cuaderno desde afuera del curso, puede tomarlo en línea haciendo clic [aquí](http://cocl.us/DP0701EN_Coursera_Week2\_LAB1).


<hr>
Copyright &copy; 2018 [Cognitive Class](https://cognitiveclass.ai/?utm_source=bducopyrightlink&utm_medium=dswb&utm_campaign=bdu). This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license/).
